File: python3/tfwconnector/serialization.py
# Copyright (C) 2018 Avatao.com Innovative Learning Kft.
# All Rights Reserved. See LICENSE file for details.
"""
TFW JSON message format

message:
{
    "key":     string,    # addressing
    "data":    {...},     # payload
    "trigger": string     # FSM trigger
}

ZeroMQ's sub-pub sockets use enveloped messages
(http://zguide.zeromq.org/page:all#Pub-Sub-Message-Envelopes)
and TFW also uses them internally. This means that on ZMQ sockets
we always send the messages key separately and then the actual
message (which contains the key as well) like so:

socket.send_multipart([message['key'], message])

The purpose of this module is abstracting away this low level behaviour.
"""

import json


def serialize_tfw_msg(message):
    """
    Create TFW multipart data from message dict
    """
    return _serialize_all(message['key'], message)


def _serialize_all(*args):
    return tuple(_serialize_single(arg) for arg in args)


def _serialize_single(data):
    """
    Return input as bytes
    (serialize input if it is JSON)
    """
    if not isinstance(data, str):
        data = json.dumps(data)
    return _encode_if_needed(data)


def _encode_if_needed(value):
    """
    Return input as bytes
    (encode if input is string)
    """
    if isinstance(value, str):
        value = value.encode('utf-8')
    return value

File: cpp/tfwconnector/serialization.cpp
/* 
 * Copyright (C) 2018 Avatao.com Innovative Learning Kft.
 * All Rights Reserved. See LICENSE file for details.
 */

#include "serialization.hpp"

#include <stdexcept>
#include <boost/property_tree/json_parser.hpp>

std::vector<std::string> tfwconnector::serialize_TFW_message(ptree message) {
    std::string key;
    /* Check if key field exists. */
    try {
        key = message.get<std::string>("key");
    } catch(...) {
        throw std::runtime_error("Invalid message format!");
    }

    /* Convert ptree to json. */
    std::ostringstream oss;
    boost::property_tree::write_json(oss, message);
    std::string json_message = oss.str();

    /* Add message parts to the vector. */
    std::vector<std::string> message_parts;
    message_parts.push_back(key);
    message_parts.push_back(json_message);

    return message_parts;
}